File: SearchEngine/SearchEngine.js
const { Tokenizer } = require("./Tokenizer");
const { Trie } = require("./Trie");
const { InvertedIndex } = require("./InvertedIndex");
const { Ranker } = require("./Ranker");

class SearchEngine {
  constructor() {
    this.tokenizer = new Tokenizer();
    this.trie = new Trie();
    this.index = new InvertedIndex();
    this.ranker = new Ranker(this.index);

    this.documents = {};  // docId → raw text
  }

  addDocument(docId, text) {
    if (!docId || !text) return;

    this.documents[docId] = text;

    const tokens = this.tokenizer.tokenize(text);

    // add to inverted index
    this.index.addDocument(docId, tokens);

    // add tokens to Trie for autocomplete
    for (const token of tokens) {
      this.trie.insert(token, docId);
    }
  }

  search(query, options = {}) {
    const tokens = this.tokenizer.tokenize(query);

    const rankedDocs = this.ranker.score(tokens).slice(0, options.limit || 10);

    return rankedDocs.map(docId => ({
      docId,
      text: this.documents[docId]
    }));
  }

  autocomplete(prefix, limit = 5) {
    const docs = this.trie.suggest(prefix.toLowerCase(), limit);
    return docs.map(docId => this.documents[docId]);
  }
}

module.exports = { SearchEngine };

File: Go/SearchEngine-go/search_engine.go
package main

type SearchResult struct {
	DocID string
	Text  string
}

type SearchEngine struct {
	tokenizer *Tokenizer
	trie      *Trie
	index     *InvertedIndex
	ranker    *Ranker
	documents map[string]string
}

func NewSearchEngine() *SearchEngine {
	index := NewInvertedIndex()
	return &SearchEngine{
		tokenizer: NewTokenizer(nil, 2),
		trie:      NewTrie(),
		index:     index,
		ranker:    NewRanker(index),
		documents: make(map[string]string),
	}
}

func (s *SearchEngine) AddDocument(docID, text string) {
	if docID == "" || text == "" {
		return
	}
	s.documents[docID] = text
	tokens := s.tokenizer.Tokenize(text)
	s.index.AddDocument(docID, tokens)
	for _, token := range tokens {
		s.trie.Insert(token, docID)
	}
}

func (s *SearchEngine) Search(query string, limit int) []SearchResult {
	if limit == 0 {
		limit = 10
	}
	tokens := s.tokenizer.Tokenize(query)
	rankedDocs := s.ranker.Score(tokens)
	if len(rankedDocs) > limit {
		rankedDocs = rankedDocs[:limit]
	}
	var results []SearchResult
	for _, docID := range rankedDocs {
		results = append(results, SearchResult{DocID: docID, Text: s.documents[docID]})
	}
	return results
}

func (s *SearchEngine) Autocomplete(prefix string, limit int) []string {
	if limit == 0 {
		limit = 5
	}
	docs := s.trie.Suggest(prefix, limit)
	var texts []string
	for _, docID := range docs {
		texts = append(texts, s.documents[docID])
	}
	return texts
}
